src/tools/ip/quicip.cpp

```diff
@@ -0,0 +1,427 @@
+        TryGetValue(argc, argv, "key_file", &KeyFile)) {
+        SecurityConfig = GetSecConfigForFile(MsQuic, Registration, KeyFile, Cert);
+        if (SecurityConfig == nullptr) {
+            printf("Failed to load certificate from file!\n");
+            return;
+        }
+    } else {
+        printf("Must specify 'selfsign', '-cert_hash' or 'cert_file'!\n");
+        return;
+    }
+
+    if (QUIC_FAILED(Status = MsQuic->ListenerOpen(Session, ServerListenerCallback, nullptr, &Listener))) {
+        printf("ListenerOpen failed, 0x%x!\n", Status);
+        goto Error;
+    }
+
+    if (QUIC_FAILED(Status = MsQuic->ListenerStart(Listener, &Address))) {
+        printf("ListenerStart failed, 0x%x!\n", Status);
+        goto Error;
+    }
+
+    printf("Press Enter to exit.\n\n");
+    getchar();
+
+Error:
+
+    if (Listener != nullptr) {
+        MsQuic->ListenerClose(Listener);
+    }
+    MsQuic->SecConfigDelete(SecurityConfig);
+    if (SelfSignedCertParams) {
+        QuicPlatFreeSelfSignedCert(SelfSignedCertParams);
+    }
+}
+
+QUIC_ADDR PublicIp;
+
+QUIC_STATUS
+QUIC_API
+ClientStreamCallback(
+    _In_ HQUIC Stream,
+    _In_opt_ void* /* Context */,
+    _Inout_ QUIC_STREAM_EVENT* Event
+    )
+{
+    switch (Event->Type) {
+    case QUIC_STREAM_EVENT_RECEIVE:
+        if (Event->RECEIVE.AbsoluteOffset + Event->RECEIVE.TotalBufferLength <= sizeof(QUIC_ADDR)) {
+            uint64_t Offset = Event->RECEIVE.AbsoluteOffset;
+            for (uint32_t i = 0; i < Event->RECEIVE.BufferCount; ++i) {
+                memcpy(
+                    ((uint8_t*)&PublicIp) + Offset,
+                    Event->RECEIVE.Buffers[i].Buffer,
+                    Event->RECEIVE.Buffers[i].Length);
+                Offset += Event->RECEIVE.Buffers[i].Length;
+            }
+        }
+        break;
+    case QUIC_STREAM_EVENT_PEER_SEND_SHUTDOWN: {
+        QUIC_ADDR_STR AddrStr = { 0 };
+        QuicAddrToString(&PublicIp, &AddrStr);
+        printf("Public IP: %s\n\n", AddrStr.Address);
+        MsQuic->ConnectionShutdown(Stream, QUIC_CONNECTION_SHUTDOWN_FLAG_NONE, 0);
+        break;
+    }
+    case QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE:
+        MsQuic->StreamClose(Stream);
+        break;
+    default:
+        break;
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+QUIC_API
+ClientConnectionCallback(
+    _In_ HQUIC Connection,
+    _In_opt_ void* /* Context */,
+    _Inout_ QUIC_CONNECTION_EVENT* Event
+    )
+{
+    switch (Event->Type) {
+    case QUIC_CONNECTION_EVENT_CONNECTED:
+        break;
+    case QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_TRANSPORT:
+    case QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_PEER:
+        break;
+    case QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE:
+        MsQuic->ConnectionClose(Connection);
+        break;
+    case QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED:
+        MsQuic->SetCallbackHandler(Event->PEER_STREAM_STARTED.Stream, ClientStreamCallback, nullptr);
```

Comment:
need (void*) cast on the ClientStreamCallback.

Reply:
Done.